src/NSubstitute/Core/CallCollection.cs

```diff
@@ -38,28 +49,47 @@ public IEnumerable<ICall> AllCalls()
 
         public void Clear()
         {
-            //Queue doesn't have a Clear method.
-            _callWrappers = new ConcurrentQueue<CallWrapper>();
+            // Queue doesn't have a Clear method.
+            _callWrappers = new ConcurrentQueue<IReceivedCallEntry>();
+        }
+ 
+        /// <summary>
+        /// Performance optimization. Allows to mark call as deleted without allocating extra wrapper.
+        /// To play safely, we track ownership, so object can be re-used only once.
+        /// </summary>
+        internal interface IReceivedCallEntry
+        {
+            ICall Call { get; }
+            bool IsDeleted { get; set; }
+            bool TryTakeEntryOwnership();
         }
 
         /// <summary>
         /// Wrapper to track that particular entry was deleted.
         /// That is needed because concurrent collections don't have a Delete method.
-        /// We null the hold value to not leak memory.
+        /// Notice, in most cases the original <see cref="Call"/> instance will be used as a wrapper itself.
         /// </summary>
-        private class CallWrapper
+        private class ReceivedCallEntry : IReceivedCallEntry
         {
             public ICall Call { get; private set; }
-            public bool IsDeleted => Call == null;
 
-            public CallWrapper(ICall call)
+            public bool IsDeleted
             {
-                Call = call;
+                get => Call == null;
+                set
+                {
+                    if (value) throw new SubstituteInternalException("Call resurrection is not expected.");
```

Comment:
If we can't set this to true, maybe make this `get` only, and keep `Delete()` to set it to false? (i.e. remove ability to attempt to set it to true using compiler)

Reply:
Done.